### ISSUE TITLE
Enable cross-compiling to ARM architectures

### ISSUE THREAD

**Issue by rafaelsamenezes** [esbmc/esbmc#516]
Using a raspberry to make ARM64 builds is slow, we would benefit from cross-compiling to it.

The difficult is that ESBMC cannot be cross-compiled due to the usage of the c2goto binary to generate the models. This can be solved by [using QEMU to run the cross-compiled c2goto](https://azeria-labs.com/arm-on-x86-qemu-user/).

**Comment by fbrausse** [comment]
Isn't the `goto` language ABI agnostic? Why not compile c2goto on the host and run it? This would remove the dependency on qemu.

**Comment by rafaelsamenezes** [comment]
The language itself is. However, we use the host libc headers during c2goto conversion. This can create slightly different versions of the models.

I wouldn't mind if this don't break the regression when testing the regression at an arm sysmtem though

**Comment by fbrausse** [comment]
I'll check whether the target's libc headers can be used by c2goto. This seems related to #496 as well.

**Comment by rafaelsamenezes** [comment]
@fbrausse can you look into this?

**Comment by fbrausse** [comment]
This branch contains 2 fixes and a [sample toolchain file](https://github.com/fbrausse/esbmc/blob/fb/x-aarch64/scripts/cmake/toolchains/aarch64-linux-gnu.cmake) for aarch64-linux-gnu, which works on my system: https://github.com/fbrausse/esbmc/tree/fb/x-aarch64
For now it uses the qemu method you mentioned earlier.

Does that work for you?

**Comment by fbrausse** [comment]
cmake would be invoked like so:
```
mkdir aarch64 && cd aarch64 && cmake .. -DCMAKE_TOOLCHAIN_FILE=../scripts/cmake/toolchains/aarch64-linux-gnu.cmake
```

**Comment by rafaelsamenezes** [comment]
> Does that work for you?

Yes, that looks great. I would suggest also trying vcpkg as it would automate the build process for other dependencies (boost). 

Is there another approach that I am not familiar?

**Comment by fbrausse** [comment]
I have zero experience with vcpkg or other non-distribution C/C++ package managers, sorry.

**Comment by fbrausse** [comment]
From https://github.com/esbmc/esbmc/issues/535#issuecomment-963656830:
> > About the clib{32,64}{,_fp} GOTO libs: they should be bundled the same way as the headers, right?
> 
> Yes

https://github.com/fbrausse/esbmc/issues/1 describes another GOTO libs issue that has come up in light of the efforts of supporting CHERI-C. My suggestion is to make bundling the GOTO libs optional and implement a fallback when they are not available. If we do that, this issue could be resolved by just not bundling the libs. This might even make QEMU dispensable in the cross-compilation to ARM.

**Comment by rafaelsamenezes** [comment]
Hm, I don't mind making it optional for the build process. But we still should have a way to bundle it if needed

I would not say I like bundling every supported goto model inside ESBMC. This will cause significant maintenance issues in future, especially now that we are extending ESBMC for multiple languages and architectures.

If you wish, you can skip bundling them inside ESBMC and just send the models (similar to what we used to do for C++) or in GOTO directly.

There are a couple of issues that need to be handled first:
1. Currently, the C-flow is hard to change.
2. We have some issues in GOTO serialization/unserialization.

**Comment by fbrausse** [comment]
We seem to agree on the general direction. Being relatively new to the project, I have a few questions:

> (similar to what we used to do for C++)

What was that?

> There are a couple of issues that need to be handled first:
>
> 1.  Currently, the C-flow is hard to change.
>  2.  We have some issues in GOTO serialization/unserialization.

I am not familiar with those issues. Could you give me some pointers?

**Comment by rafaelsamenezes** [comment]
> What was that?

We don't ship the models directly in the build. The user has to include some of the headers with the operational models: https://github.com/esbmc/esbmc/tree/master/src/cpp/library

> I am not familiar with those issues. Could you give me some pointers?

1. The build/execution process of esbmc assumes that the libc models are going to be shipped directly. 
2. The serialization/unserialization is that esbmc (afaik) cannot run .goto files directly

**Comment by fbrausse** [comment]
> We don't ship the models directly in the build.

> 1. The build/execution process of esbmc assumes that the libc models are going to be shipped directly.

Both should be taken care of now by #581: the sources in c2goto/library are bundled and translated in case the GOTO-libs are missing.

This is slow, though, so as an extension it would probably be beneficial if ESBMC could construct these only once and cache the translations in the GOTO format. That would probably need to be tied to the specific ESBMC binary, e.g., ```esbmc-md5`md5sum $ESBMC_binary | cut -d' ' -f1`/``` in `~/.cache` or `/tmp`.

> We have some issues in GOTO serialization/unserialization.
> The serialization/unserialization is that esbmc (afaik) cannot run .goto files directly

Is this really required to store/load the compiled GOTO libs? It does work already for those created during the build...

**Comment by rafaelsamenezes** [comment]
> That would probably need to be tied to the specific ESBMC binary

I didn't understand your idea, but shouldn't c2goto be used?

>  It does work already for those created during the build

Yes, .goto generated through c2goto works and generates from esbmc I couldn't run. It may be just a frontend issue. I haven't tested it again.

**Comment by fbrausse** [comment]
> I didn't understand your idea, but shouldn't c2goto be used?

At the moment when the builtin libs are missing, c2goto is not invoked, but the libc-sources are processed directly and merged into the context: https://github.com/esbmc/esbmc/blob/f49df29331f84d7685e84b34ab6521a53ea6bc82/src/c2goto/cprover_library.cpp#L219-L222

This takes about 1 second on my laptop and the generated GOTO representation is only used for this run of esbmc. I proposed to cache this representation somewhere tied to the binary used to generate them. Another option would be to invoke c2goto, but I didn't see the benefit of doing that unless we can store the results somewhere. In any case, this is an optimization we might want to look into separately from this issue.

What is left to be done for this issue?

**Comment by fbrausse** [comment]
> However, we use the host libc headers during c2goto conversion. This can create slightly different versions of the models.

#602 takes care of that by a) esmbc's libc headers taking precedence over the system and b) both, c2goto and esbmc, supporting `--sysroot`. We would only have to set `C2GOTO_SYSROOT` appropriately in the CMake toolchain file as well.

**Comment by rafaelsamenezes** [comment]
> What is left to be done for this issue?

For me, a CI script for testing and generation of ARM artifacts

**Comment by rafaelsamenezes** [comment]
This is no longer valid as we can get native arm builds from the CI.
